core/slice.go

```diff
@@ -246,10 +246,11 @@ func (sl *Slice) relayPh(block *types.Block, appendTime *time.Duration, reorg bo
 	nodeCtx := common.NodeLocation.Context()
 
 	if nodeCtx == common.ZONE_CTX {
-		sl.phCachemu.RLock()
+		sl.phCachemu.Lock()
```

Comment:
we should not do this because, you are locking append here during the generation of the pending header with the udpated state root

Reply:
This is necessary. It is nonsensical to allow bestPh to change after we read it on L253, but before we use it later.

e.g. `bestPh` is read at L253, again at L268, and internally in `writeToPhCacheAndPickPhHead(...)`. It would be nonsense if `writeToPhCacheAndPickPhHead()` uses one bestPh, but then an async thread changes the bestPh to something random before we read bestPh back on L255.

This entire function needs to assume control of the PH cache during its execution, hence my change to the phcache lock

Reply:
might be, but we need to think of a better solution in this case, because whole point of generating state root after append is to not lock append, and this violates that

Reply:
why would it be nonsense ? for the write to use different bestPh, because update and pick are separate processes

Reply:
Also this is only in the zone case and miner should always mine on bestPh

Reply:
I don't understand what you are saying. The original code also locked during generation of the pending header. All I did was move the lock to the top of the function instead of right next to the generation call. Either way it still locks for the duration of generation.